enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,1091 @@
+* Defining updateability for non-ostree systems is left to a future enhancement
+* Protecting against data corruption - we rely on file system 
+  to maintain the backed up file integrity
+
+## Proposal
+
+### Workflow Description
+
+**MicroShift administrator** is a human responsible for preparing
+ostree commits and scheduling devices to use these commits.
+
+Upgrade:
+
+1. MicroShift administrator prepares a new ostree commit
+1. MicroShift administrator schedules device to reboot and use new ostree commit
+1. Device boots to new commit
+1. Operating System, greenboot, and MicroShift take actions without any additional intervention
+
+Manual rollback:
+
+1. MicroShift administrator instructs MicroShift to do a restore on next boot
+1. MicroShift administrator stages an ostree deployment with MicroShift
+   that was already running and reboots the device
+1. Staged ostree deployment boots
+1. MicroShift will restore the backup matching current ostree deployment
+1. MicroShift will run.
+
+### API Extensions
+
+Metadata persisted on filesystem related to the functionality described 
+in this enhancement such as version metadata, next boot action, and any other,
+is considered internal implementation detail and not an API to be consumed.
+Content (schema) and location of these files are subject to change.
+
+### Implementation Details/Notes/Constraints [optional]
+
+### Risks and Mitigations
+
+Being a GA feature from the beginning the risks are not foreseeing fail scenarios in advance and implementation bugs
+that are not caught and fixed through graduation process.
+
+To mitigate the risks, a thorough review of the enhancement must be done by MicroShift, OpenShift, and RHEL teams,
+and making sure testing strategy is sound and prioritized equally with the feature development.
+
+### Drawbacks
+
+N/A
+
+## Design Details
+
+### Definitions
+
+- **ostree commit**: TODO
+- **ostree deployment**: TODO (Remember, new deployment might include K8s apps not previously running)
+- **Rollback**: booting older (that already ran on the device) ostree deployment - either due to greenboot or manual intervention
+- **Backup**: backing up `/var/lib/microshift`
+- **Restore**: restoring `/var/lib/microshift`
+- **Data migration** - procedure of transitioning MicroShift's data to be compatible with newer binary
+- **Version metadata**: File residing in MicroShift data dir containing version of MicroShift and ID of ostree deployment
+- **MicroShift greenboot healthcheck**: Program verifying the status of MicroShift's cluster
+
+### Preface
+
+Every action related to procedure described in this enhancement is 
+performed after system's boot rather than immediately before shutdown. 
+Greenboot's healthchecks, green and red scripts are executed independent of MicroShift's processes.
+Actions related to backup, restore, and data migration may be performed with MicroShift 
+partially running, i.e. only etcd and kube-apiserver are running.
```

Comment:
We cannot have `etcd` or apiserver running while performing db operations, or we are data inconsistency.
If we start microshift exe, all the db manipulations must be over before we attempt starting `etcd` and other high-level services.